src/Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.cpp

```diff
@@ -0,0 +1,152 @@
+#include <Storages/MergeTree/ReplicatedMergeTreeLogEntry.h>
+
+
+#include <common/types.h>
+#include <optional>
+#include <mutex>
+#include <city.h>
+#include <algorithm>
+#include <atomic>
+
+namespace DB
+{
+
+// minimum interval (seconds) between checks if chosen replica finished the merge.
+static const auto RECHECK_MERGE_READYNESS_INTERVAL_SECONDS = 1;
+
+// don't refresh state too often (to limit number of zookeeper ops)
+static const auto REFRESH_STATE_MINIMUM_INTERVAL_SECONDS = 5;
+
+// refresh the state automatically it it was not refreshed for a longer time
+static const auto REFRESH_STATE_MAXIMUM_INTERVAL_SECONDS = 30;
+
+
+ReplicatedMergeTreeMergeStrategyPicker::ReplicatedMergeTreeMergeStrategyPicker(StorageReplicatedMergeTree & storage_)
+    : storage(storage_)
+{}
+
+
+bool ReplicatedMergeTreeMergeStrategyPicker::isMergeFinishedByReplica(const String & replica, const ReplicatedMergeTreeLogEntryData & entry)
+{
+    // those have only seconds resolution, so recheck period is quite rough
+    auto reference_timestamp = entry.last_postpone_time;
+    if (reference_timestamp == 0)
+        reference_timestamp = entry.create_time;
+
+    // we don't want to check zookeeper too frequent
+    if (time(nullptr) - reference_timestamp >= RECHECK_MERGE_READYNESS_INTERVAL_SECONDS)
+    {
+        return storage.checkReplicaHavePart(replica, entry.new_part_name);
+    }
+
+    return false;
+}
+
+
+// that will return the same replica name for ReplicatedMergeTreeLogEntry on all the replicas (if the replica set is the same).
+// that way each replica knows who is responsible for doing a certain merge.
+
+// in some corner cases (added / removed / deactivated replica)
+// nodes can pick different replicas to execute merge and wait for it (or to execute the same merge together)
+// but that doesn't have a significant impact (in one case it will wait for the execute_merges_on_single_replica_time_threshold,
+// in another just 2 replicas will do the merge)
+std::optional<std::string> ReplicatedMergeTreeMergeStrategyPicker::pickReplicaToExecuteMerge(const ReplicatedMergeTreeLogEntryData & entry)
+{
+    time_t threshold = execute_merges_on_single_replica_time_threshold;
+    if (
+        threshold == 0 // feature turned off
+        || entry.type != ReplicatedMergeTreeLogEntry::MERGE_PARTS // it is not a merge log entry
+       )
+        return std::nullopt;
+
+    auto now = time(nullptr);
+
+    if (entry.create_time + threshold <= now) // too much time waited
+        return std::nullopt;
+
+    if (now - last_refresh_time > REFRESH_STATE_MAXIMUM_INTERVAL_SECONDS) // last state refresh was too long ago, need to sync up the replicas list
+        refreshState();
+
+    auto hash = getEntryHash(entry);
+
+    std::lock_guard lock(mutex);
+
+    auto num_replicas = active_replicas.size();
+
+    if (num_replicas==0)
+        return std::nullopt;
+
+    auto replica_index = static_cast<int>(hash % num_replicas);
```

Comment:
Seems like this code very unreliable. What if some replica has fewer replicas in it's cache than current?

Reply:
The worst scenario - it will pick some replica which will go offline immediately. So it will wait for it till the threshold will be reached (after that either will do merge by it's own) or till cache refresh will happen.

I don't think we can provide better guarantees here (well at least with compact efforts).

Normally it should 'just work' . In corner cases it can either do the merge on multiple replicas simultaniously, either the merge will be executed with the delay (up to defined threshold)